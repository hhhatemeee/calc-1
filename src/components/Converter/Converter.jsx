import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import ScreenOther from '../ScreenOther/ScreenOther';
import { ReactComponent as Loader } from '../../img/Loader.svg';
import ConverterSwitch from './ConverterSwitch/ConverterSwitch';
import { CURRENCY_MOCK } from '../../variables';
import ConvertationService from '../../services/convertationService';
import KeyBoardBasic from '../KeyBoardBasic/KeyBoardBasic';
import ModalInfo from './ModalInfo/ModalInfo';

import './Converter.scss';

const Converter = (props) => {
  useEffect(() => { }, [props.currentService])

  const [from, setFrom] = useState(
    {
      name: CURRENCY_MOCK.RUB[0],
      value: CURRENCY_MOCK.RUB[1],
    });
  const [to, setTo] = useState(
    {
      name: CURRENCY_MOCK.USD[0],
      value: CURRENCY_MOCK.USD[1],
    });
  const setOptions = () => props.services.filter((el) => !props.listLimit.includes(el.name));

  const setFromValue = (value) => setFrom(value);
  const setToValue = (value) => setTo(value);

  const onClick = () => {
    props.setLoading(true);
    props.setFetching(true);
    setCurrentCourse(to.name);
  }

  const setCurrentCourse = async (e) => {
    //If an event came from the selector, then process it.
    props.setFetching(true);
    if (e.target && e.target.value) {
      const cc = await ConvertationService.getConvertation(e.target.value);
      await props.setCurrentCourse(cc)
      return;
    }

    //If the name of the currency has come
    const cc = await ConvertationService.getConvertation(e);
    await props.setCurrentCourse(cc);
  }

  const onChange = (e) => {
    const targetValue = e && e.target && e.target.value
      ? e.target.value
      : 'CC';

    props.setFetching(true);
    props.setCurrentService(targetValue)
    ConvertationService.switchService(targetValue);
    ConvertationService.updateCurrencyList();
  }

  const onSwapCurrency = () => {
    const currentCurrencies = { from: { ...from }, to: { ...to } };

    setFrom(currentCurrencies.to);
    setTo(currentCurrencies.from);
    props.handleBasicCurrency(to.name);
    props.setCurrentCourse(1 / props.currentCourse);
  }

  return (
    <div className='converter__container'>
      <ModalInfo servicesStatus={props.servicesStatus} servicesUrl={props.servicesUrl} />
      <ScreenOther
        CURRENCY_TABLE={props.CURRENCY_TABLE}
        currencyList={props.currencyList}
        setFromValue={setFromValue}
        setToValue={setToValue}
        to={to}
        from={from}
        currentNumber={props.currentNumber}
        setCurrentCourse={setCurrentCourse}
        resultNumber={props.resultNumber}
        fontSizeOne={props.fontSizeOne}
        fontSizeTwo={props.fontSizeTwo}
      />
      <i className='ico-Swap converter__swap-button' onClick={onSwapCurrency} data-testid='swapBtn' />
      <span onClick={onClick} className='converter__update'>
        Update rates
        {props.isLoading && <Loader className='loader' />}
      </span>
      <ConverterSwitch
        listLimit={props.listLimit}
        currentService={props.currentService}
        onChange={onChange}
        options={setOptions()}
      />
      <KeyBoardBasic buttons={props.buttons} handleCurNum={props.handleCurNum} />
    </div>
  )
}

Converter.propTypes = {
  currencyList: PropTypes.array,
  CURRENCY_TABLE: PropTypes.object,
  resultNumber: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  currentNumber: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  fontSizeOne: PropTypes.number,
  fontSizeTwo: PropTypes.number,
  isLoading: PropTypes.bool,
  buttons: PropTypes.array,
  handleCurNum: PropTypes.func.isRequired,
  setLoading: PropTypes.func,
  setCurrentCourse: PropTypes.func,
  services: PropTypes.array,
  listLimit: PropTypes.array,
  currentService: PropTypes.string,
  setCurrentService: PropTypes.func,
  servicesStatus: PropTypes.object,
};

Converter.defaultProps = {
  currencyList: [],
  CURRENCY_TABLE: {},
  resultNumber: '0',
  currentNumber: '0',
  fontSizeOne: 88,
  fontSizeTwo: 88,
  isLoading: false,
  buttons: [],
  services: [],
  listLimit: [],
  currentService: 'CC',
  servicesStatus: {},
  handleCurNum: () => console.log('Не определена функция handleCurNum'),
  handleBasicCurrency: () => console.log('Не определена функция handleBasicCurrency'),
  updateCurrencyList: () => console.log('Не определена функция updateCurrencyList'),
  switchService: () => console.log('Не определена функция switchService'),
  setLoading: () => console.log('Не определена функция setLoading'),
  setCurrentCourse: () => console.log('Не определена функция setCurrentCourse'),
  handleConvertaionCurrency: () => console.log('Не определена функция handleConvertaionCurrency'),
  setCurrentService: () => console.log('Не определена функция setCurrentService'),
};

export default Converter;