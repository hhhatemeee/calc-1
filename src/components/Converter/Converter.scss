@import "../../scss/variables.scss";
@import "../../scss/icons.scss";
@import "../../scss/fonts.scss";


.converter__container{
  margin: 40px 20px;
 
}

.converter__update{
  margin-left: 4px;
  text-decoration: underline;
  cursor: pointer;
  transition: 0.1s ease-in;
  user-select: none;
  display: flex;

  &:hover{
    color: rgba($color-black,0.8);
  }

  &:active{
    color: rgba($color-black,0.5);
  }
}
.converter__switch{
  display: flex;
  align-items: center;
  margin-left: 4px;
  white-space: pre;
  user-select: none;
  transition: 0.1s ease-in;

  .select{
    display: flex;
    width: 100%;
  }

  p{
    width: 125px;
    padding: 0;
    margin: 0;
  }

  .convertner__current-service{
    font-weight: 500;
    font-size: 20px;
  }
}

.loader{
  width: 20px;
  height: 20px;
  // margin: 0 5px !important;
  left: 130px;
  position: absolute;
}