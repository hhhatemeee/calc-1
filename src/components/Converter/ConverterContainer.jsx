import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types';

import splittingNumber from '../../helpers/splittingNumber';
import { setCurrentCourseCreator, setCurrentServiceCreator, setFetchingCreator, setLoadingCreator } from '../../redux/convertationReducer';
import Converter from './Converter'
import { BUTTONS_MOCK } from '../../variables';
import { KEYS_NAME } from '../../variables';
import ConvertationService from '../../services/convertationService';


const ConverterContainer = (props) => {
  const options = [];
  const [currentNumber, setCurrentNumber] = useState(0);
  const [resultNumber, setResultNumber] = useState(0);
  const [fontSizeOne, setFontSizeOne] = useState(88);
  const [fontSizeTwo, setFontSizeTwo] = useState(88);
  const [servicesStatusApi, setServicesStatusApi] = useState({});
  const [currencyList, setCurrencyList] = useState([]);
  const CURRENCY_TABLE = {
    RUB: '₽',
    EUR: '€',
    ALL: 'L',
    FKP: '£',
    GIP: '£',
    IRR: '﷼',
    JMD: '$',
    AUD: '$',
    LAK: '₭',
  };

  useEffect(() => {
    setServicesStatusApi(ConvertationService.getStatusApi());
  }, [props.isFetching]);

  const onKeyDown = (e) => {
    if ((e.key >= '0' && e.key <= '9') || e.key === KEYS_NAME.Backspace || e.key === '.') {
      handleCurNum(e.key);
    }

    if (e.ctrlKey && e.keyCode === 86) {
      navigator.clipboard.readText()
        .then(text => {
          if (!Number.isNaN(Number(text)))
            handleCurNum(text);
        })
    }
  };

  useEffect(() => generateList(), [props.currencyList]);

  useEffect(() => onKeyDown(props.currentKey), [props.currentKey]);

  useEffect(() => {
    let curNum = currentNumber.toString();

    if (props.currentCourse > 1) {
      setResultNumber((Number(curNum) / props.currentCourse).toFixed(2));
      return;
    }
    setResultNumber((Number(curNum) / props.currentCourse).toFixed(2));
  }, [props.currentCourse]);

  /**
   * Calculates the font size for the display
   * @param {number} num - Length of the current number
   * @param {boolean} isSecondLine - If true, then calculate the length for the second row
   * @returns 
   */
  const getFontSize = (num, isSecondLine) => {
    let size;
    size = 88;

    if (isSecondLine) {
      if (num > 6) {
        size = -(num * 7.5) + 126.74;

      }

      if (num >= 10) {
        size = -(num * 4) + 89.2;
      }

      if (num > 13) {
        size = -(num * 2.8) + 75.4;
      }
      setFontSizeTwo(size);

      return;
    }

    if (num > 6) {
      size = -(num * 8.2) + 136.6;
    }

    if (num > 10) {
      size = -(num * 3.08) + 81.88;
    }

    if (num > 15) {
      size = -(num * 1.48) + 57.85;
    }

    setFontSizeOne(size)
  }

  /**
   * Handler for the current button click
   * @param {string/number} button -  the current value of the button
   * @returns 
   */
  const handleCurNum = (button) => {
    if (button.toString().length < 17) {

      let curNum = currentNumber.toString();
      let value = button;
      switch (value) {
        case BUTTONS_MOCK.reset:
          curNum = '0';
          break;
        case BUTTONS_MOCK.delete:
          curNum = curNum.toString().slice(0, curNum.length - 1)
          break;
        case KEYS_NAME.Backspace:
          curNum = curNum.toString().slice(0, curNum.length - 1)
          break;
        case '.':
          if (curNum.includes('.')) {
            return;
          }
          value = '.';
          curNum += value;
          break;
        case BUTTONS_MOCK.dot:
          if (curNum.includes('.')) {
            return;
          }
          value = '.';
          curNum += value;
          break;
        default:
          if (curNum.length === 15) {
            return
          }
          value = button;
          console.log(value, curNum);
          curNum += value;
          break;
      }

      //If the last value of the number is a point, then save
      if (curNum.split('').at(-1) === '.') {
        setCurrentNumber(curNum);
        return;
      }

      // You cannot enter more than two numbers after the dot
      if (curNum.includes('.') && curNum.split('.')[1].length > 2) {
        return;
      }

      getFontSize(splittingNumber(curNum).length);

      setCurrentNumber(Number(curNum));

      // If the rate is greater than 1 then divide the numbers
      if (props.currentCourse > 1) {
        const result = (Number(curNum) / props.currentCourse).toFixed(2);
        getFontSize(result.toString().length, true);

        setResultNumber(result);
        return;
      }

      const result = (Number(curNum) * props.currentCourse).toFixed(2);
      getFontSize(result.toString().length, true);
      setResultNumber(result);
    }
  }

  // Generate a list of currencies
  const generateList = () => {
    props.currencyList.forEach((val) => {
      options.push({ name: val, value: val })
    })

    setCurrencyList(options);
  }

  return (
    <Converter
      onKeyDown={onKeyDown}
      CURRENCY_TABLE={CURRENCY_TABLE}
      currentNumber={currentNumber}
      currencyList={currencyList}
      buttons={props.buttons}
      handleCurNum={handleCurNum}
      isLoading={props.isLoading}
      setLoading={props.setLoading}
      setCurrentCourse={props.setCurrentCourse}
      resultNumber={resultNumber}
      fontSizeOne={fontSizeOne}
      fontSizeTwo={fontSizeTwo}
      services={props.services}
      listLimit={props.listLimit}
      currentService={props.currentService}
      setCurrentService={props.setCurrentService}
      servicesStatus={servicesStatusApi}
      servicesUrl={props.servicesUrl}
      setFetching={props.setFetching}
    />
  )
}

const mapStateToProps = (state) => {
  return {
    currencyList: state.convertation.currencyList,
    buttons: state.buttonList,
    isLoading: state.convertation.isLoading,
    currentCourse: state.convertation.currentCourse,
    services: state.convertation.services,
    servicesUrl: state.convertation.servicesUrl,
    currentService: state.convertation.currentService,
    isFetching: state.convertation.isFetching,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setLoading: (isLoading) => dispatch(setLoadingCreator(isLoading)),
    setCurrentCourse: (value) => dispatch(setCurrentCourseCreator(value)),
    setCurrentService: (service) => dispatch(setCurrentServiceCreator(service)),
    setFetching: (bool) => dispatch(setFetchingCreator(bool)),
  }
}

ConverterContainer.propTypes = {
  currencyList: PropTypes.array,
  buttons: PropTypes.array,
  isLoading: PropTypes.bool,
  currentCourse: PropTypes.number,
  setLoading: PropTypes.func,
  setCurrentCourse: PropTypes.func,
  services: PropTypes.array,
  listLimit: PropTypes.array,
  currentService: PropTypes.string,
  setCurrentService: PropTypes.func,
  currentKey: PropTypes.object
};

ConverterContainer.defaultProps = {
  currencyList: [],
  buttons: [],
  isLoading: false,
  currentCourse: 1,
  services: [],
  listLimit: [],
  currentService: 'CC',
  currentKey: {},
  setLoading: () => console.log('Не определена функция setLoading'),
  setCurrentCourse: () => console.log('Не определена функция setCurrentCourse'),
  setCurrentService: () => console.log('Не определена функция setCurrentService'),
};

export default connect(mapStateToProps, mapDispatchToProps)(ConverterContainer);
