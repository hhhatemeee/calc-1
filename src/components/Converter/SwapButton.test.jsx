import { render, screen, fireEvent } from '@testing-library/react';
import { toMatchDiffSnapshot } from 'snapshot-diff';

import Converter from './Converter';

expect.extend({ toMatchDiffSnapshot });

describe('SwapButton test', () => {
  it('The presence of a button on the screen', () => {
    render(<Converter />);

    const child = screen.queryByTestId('swapBtn');

    expect(child).not.toBeNull();
  });


  it('check the onClick the swapButton', () => {
    const converter = render(<Converter />);
    const child = screen.queryByTestId('swapBtn');

    fireEvent.click(child)
    const converterBeforeClick = render(<Converter />);

    expect(converter).toMatchDiffSnapshot(converterBeforeClick);
  })
});
