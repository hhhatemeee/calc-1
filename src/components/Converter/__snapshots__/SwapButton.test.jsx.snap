// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapButton test check the onClick the swapButton 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -272,11 +272,11 @@
                0
              </p>
              <p
                class=\\"screen-line__currency\\"
              >
-               $
+               ₽
              </p>
            </span>
            <div
              class=\\"select\\"
            >
@@ -300,11 +300,11 @@
                0
              </p>
              <p
                class=\\"screen-line__currency\\"
              >
-               ₽
+               $
              </p>
            </span>
            <div
              class=\\"select\\"
            >"
`;
